GVFS/GVFS/CommandLine/DehydrateVerb.cs

```diff
@@ -145,31 +164,139 @@ protected override void Execute(GVFSEnlistment enlistment)
                 // Local cache and objects paths are required for TryDownloadGitObjects
                 this.InitializeLocalCacheAndObjectsPaths(tracer, enlistment, retryConfig, serverGVFSConfig: null, cacheServer: null);
 
-                if (this.TryBackupFiles(tracer, enlistment, backupRoot))
+                if (string.IsNullOrEmpty(this.Folders))
                 {
-                    if (this.TryDownloadGitObjects(tracer, enlistment, retryConfig) &&
-                        this.TryRecreateIndex(tracer, enlistment))
+                    this.RunFullDehydrate(tracer, enlistment, backupRoot, retryConfig);
+                }
+                else
+                {
+                    string[] folders = this.Folders.Split(new[] { ";" }, StringSplitOptions.RemoveEmptyEntries);
+
+                    if (folders.Length > 0)
                     {
-                        // Converting the src folder to partial must be the final step before mount
-                        this.PrepareSrcFolder(tracer, enlistment);
-                        this.Mount(tracer);
+                        if (cleanStatus)
+                        {
+                            this.DehydrateFolders(tracer, enlistment, backupRoot, folders);
+                        }
+                        else
+                        {
+                            this.ReportErrorAndExit("Must have a clean git status to dehydrate folders.");
+                        }
+                    }
+                    else
+                    {
+                        this.ReportErrorAndExit("No folders to dehydrate.");
+                    }
+                }
+            }
+        }
+
+        private void DehydrateFolders(JsonTracer tracer, GVFSEnlistment enlistment, string backupRoot, string[] folders)
+        {
+            List<string> foldersToDehydrate = new List<string>();
+
+            if (!this.ShowStatusWhileRunning(
+                () =>
+                {
+                    string ioError;
+                    foreach (string folder in folders)
+                    {
+                        string fullPath = Path.Combine(enlistment.WorkingDirectoryRoot, folder);
+                        string backupPath = Path.Combine(backupRoot, folder);
+                        if (this.fileSystem.DirectoryExists(fullPath))
+                        {
+                            if (!this.TryIO(tracer, () => this.fileSystem.CopyDirectoryRecursive(fullPath, backupPath), $"Backing up {folder} to {backupPath}", out ioError) ||

```

Comment:
We would then need platform specific handling here which I would like to avoid if at all possible.  Aren't the empty placeholder files just files with specific reparse data and copying should work like with any other file?  Is there an issue with moving the file with the reparse data?  Will it cause issues with explorer or windows?  I will do some testing and see if I run into anything.

Reply:
> Aren't the empty placeholder files just files with specific reparse data and copying should work like with any other file? Is there an issue with moving the file with the reparse data? Will it cause issues with explorer or windows?
> 
> I can't remember how empty placeholders will behave in this scenario (@cgallred might).
> 
> I will do some testing and see if I run into anything.
> 
> When you give this a test please be sure to include a scenario were there is a second mounted repo on the same volume while the dehydrate is being run.  My concern was that ProjFS might return an error like ERROR_FILE_SYSTEM_VIRTUALIZATION_UNAVAILABLE if you attempt to read the empty placeholder while the provider is not running.
> 
> We would then need platform specific handling here which I would like to avoid if at all possible.
> 
> That would definitely be a downside to the approach I suggested, however, if a large cone of the repo is being dehydrated the move might be significantly faster than a copy.  If the performance of the copy is not a problem (and there are no issues with copying empty placeholders) then I think we could stick with copy and keep the move option in mind if we find ourselves having to speed up this path.

Sorry for the delay, I was on vacation.  If the provider is running then copying will result in a content recall.  The file system doesn't have a "copy" operation.  Copy is an operation synthesized at the Win32 API level.  In the file system we see a copy as open-read-close for the copy source and open(possibly w/ create new)-write-close for the copy destination.  The file system does have a move/rename operation.  That should not result in a content recall.